util/tls/tls_socket.cc

```diff
@@ -468,26 +562,161 @@ void TlsSocket::CancelOnErrorCb() {
 }
 
 void TlsSocket::RegisterOnRecv(OnRecvCb cb) {
-  LOG(FATAL) << "Not implemented";
+  next_sock_->RegisterOnRecv(
+      [recv_cb = std::move(cb), this](const RecvNotification& rn) { RecvAsync(rn, recv_cb); });
 }
 
-void TlsSocket::ResetOnRecvHook() {
-  LOG(FATAL) << "Not implemented";
+void TlsSocket::RecvAsync(const RecvNotification& rn, const OnRecvCb& recv_cb) {
+  if (std::holds_alternative<FiberSocketBase::RecvCompletion>(rn.read_result) ||
+      std::holds_alternative<std::error_code>(rn.read_result)) {
+    if (recv_cb) {
+      recv_cb(rn);
+    }
+    return;
+  }
+
+  if (std::holds_alternative<io::MutableBytes>(rn.read_result)) {
+    auto& buf{std::get<io::MutableBytes>(rn.read_result)};
+    auto input_buf{engine_->PeekInputBuf()};
+    CHECK_GE(input_buf.size(), buf.size());
+    std::memcpy(input_buf.data(), buf.data(), buf.size());
+    engine_->CommitInput(buf.size());
+
+    // Loop to read and deliver all available decrypted application data.
+    // Note: app_buf is stack-allocated and only valid during this function's scope.
+    // RecvNotification's buffer must not be accessed after the callback returns.
+    // See fiber_socket_base.h:131-132 for buffer lifetime contract.
+    while (true) {
+      static constexpr size_t kAppBufSize = 4096;
+      std::array<uint8_t, kAppBufSize> app_buf{};
+      int red_res = engine_->Read(app_buf.data(), app_buf.size());
+      if (red_res > 0) {
+        RecvNotification internal_rn;
+        internal_rn.read_result = io::MutableBytes{app_buf.data(), static_cast<size_t>(red_res)};
+        recv_cb(internal_rn);
+        // Continue looping to deliver all buffered decrypted data to upstream socket.
+        continue;
+      } else if ((red_res == Engine::NEED_READ_AND_MAYBE_WRITE) ||
+                 (red_res == Engine::NEED_WRITE)) {
+        HandleOpAsync(red_res);
+        // After handling, check if more decrypted data is available and continue loop.
+        continue;
+      } else {
+        // EOF_STREAM (connection closed probably), 0 (no data read) or unexpected error
+
+        break;
+      }
+    }
+  }
 }
 
 io::Result<size_t> TlsSocket::TrySend(io::Bytes buf) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  iovec v{const_cast<uint8_t*>(buf.data()), buf.size()};
+  return TrySend(&v, 1);
 }
 
 io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  size_t total_written = 0;
+  size_t iov_idx = 0;
+  size_t iov_offset = 0;
+  while (iov_idx < len) {
+    // Flush any pending output from previous iterations or previous calls.
+    while (engine_->OutputPending() > 0) {
+      auto ec = MaybeSendOutputAsync();
```

Comment:
based on the implementation of `MaybeSendOutputAsync` I think you have a conceptual bug here.
MaybeSendOutputAsync does not guarantee that the output buffer will be eventually flushed. This will cause short writes.